### ISSUE TITLE
Can trivially improve g-zip compress ratio

### ISSUE THREAD

**Issue by qeleb** [simonwep/vite-plugin-optimize-css-modules#8]
Currently, the default dictionary is: `_-abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789`

Based on a character frequency analysis of the code in a few of my projects it should be re-ordered to this: `etionraldfps0gx-1chbum4v6w25k9y873zjHCONADLYqBEFGIJKMPQRSTUVWXZ_`

For me this measurably improved g-zip compression ratio without any cost. Perhaps a deeper analysis could be done for the most optimal default dictionary, but this is at least a step in the right direction.

**Comment by simonwep** [comment]
Wow, this is super interesting! I definitely haven't thought about that but will for sure have a look at it!

**Comment by simonwep** [comment]
I added benchmarks and ran them against your dictionary, of course this his highly skewed because of me only testing against two libraries at the moment. I could only notice a small improvement in size but a major slowdown in bundle time:

#### Current dictionary 

| Input                                                                            | Build Time                            | Gzip Size                                | Brotli Size                             |
|----------------------------------------------------------------------------------|---------------------------------------|------------------------------------------|-----------------------------------------|
| [bootstrap-5.0.2.module.css](benchmarks/fixtures/bootstrap-5.0.2.module.css)     | 525ms (_**-94.06%**_ / _**-8311ms**_) | 21.3 kB (_**-26.53%**_ / _**-7.69 kB**_) | 21.3 kB (_**-27.54%**_ / _**-6 kB**_)   |
| [materialize-1.0.0.module.css](benchmarks/fixtures/materialize-1.0.0.module.css) | 572ms (_**-92.59%**_ / _**-7156ms**_) | 20.1 kB (_**-19.70%**_ / _**-4.93 kB**_) | 20.1 kB (_**-21.33%**_ / _**-4.3 kB**_) |


#### Your dictionary

| Input                                                                            | Build Time                             | Gzip Size                                | Brotli Size                              |
|----------------------------------------------------------------------------------|----------------------------------------|------------------------------------------|------------------------------------------|
| [bootstrap-5.0.2.module.css](benchmarks/fixtures/bootstrap-5.0.2.module.css)     | 1106ms (_**-88.00%**_ / _**-8114ms**_) | 21.3 kB (_**-26.66%**_ / _**-7.73 kB**_) | 21.3 kB (_**-28.22%**_ / _**-6.15 kB**_) |
| [materialize-1.0.0.module.css](benchmarks/fixtures/materialize-1.0.0.module.css) | 1112ms (_**-87.46%**_ / _**-7751ms**_) | 20.1 kB (_**-19.80%**_ / _**-4.95 kB**_) | 20.1 kB (_**-20.64%**_ / _**-4.16 kB**_) |

Do you have any public accessible libraries/css code that I can use to test this against or references on why the order of your dictionary should have a large impact? I can imagine you ordered it based on the frequency of how often each character is used but I'd like to include some references before I make that change :D

**Comment by qeleb** [comment]
I ran a character frequency in my particular project’s css.. I suspect if you do that for these specific projects you’ll observe some real improvement. When I get a chance, I can look into a better general dictionary

**Comment by simonwep** [comment]
I just ran it a few more times, seems like my mac was occupied with something else - you're right, the build time is the same.

| Input | Build Time | Gzip Size | Brotli Size |
| --- | --- | --- | --- |
| [bootstrap-5.0.2.module.css](benchmarks/fixtures/bootstrap-5.0.2.module.css) | 550ms (_**-93.93%**_ / _**-8515ms**_) | 21.3 kB (_**-26.66%**_ / _**-7.73 kB**_) | 21.3 kB (_**-28.22%**_ / _**-6.15 kB**_) |
| [materialize-1.0.0.module.css](benchmarks/fixtures/materialize-1.0.0.module.css) | 569ms (_**-93.29%**_ / _**-7917ms**_) | 20.1 kB (_**-19.80%**_ / _**-4.95 kB**_) | 20.1 kB (_**-20.64%**_ / _**-4.16 kB**_) |

I also thought about character frequency, but this would've been needed based on the the CSS that is compiled against which should be possible in the scope of the plugin?! I'll take a closer look at this! :D

**Comment by qeleb** [comment]
this is the same question I’m asking.
If it’s done in the plugin, we’d have to iterate over the css twice (or perhaps cache some info like character frequency run to run). this would definitely bring the best results, but I think having a better general purpose dictionary may yield results nearly as good and for much easier— I think at minimum we should take that approach first.

**Comment by simonwep** [comment]
Yeah I'd also favor a more general dictionary, if you want you can have a go at this since you already investigated this for your own personal project and open an PR in case you find anything :)
